### ISSUE TITLE
`clippy::manual_async_fn` suggests editing macro definition

### ISSUE THREAD

**Issue by PatchMixolydic** [rust-lang/rust-clippy#12407]
### Summary

If a macro expands to a function that returns `impl Future` with an `async` block in its body, Clippy will suggest editing the macro definition to make it expand to an `async fn`, even if the macro also expands to synchronous functions.

### Reproducer

I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=abf34a8362c77035d47902eae0ec738c)):

```rust
macro_rules! expr_fn {
    (
        $(
            $(#[$m:meta])*
            $v:vis $(override($($overrides:tt),* $(,)?))? fn $name:ident $([$($params:tt)*])? (
                $($arg_name:ident: $arg_typ:ty),* $(,)?
            ) $(-> $ret_ty:ty)? = $e:expr;
        )*
    ) => {
        $(
            $(#[$m])*
            $v $($($overrides)*)? fn $name$(<$($params)*>)?(
                $($arg_name: $arg_typ),*
            ) $(-> $ret_ty)? {
                $e
            }
        )*
    };
}

expr_fn! {
    fn _hello() -> impl Future<Output = ()> = async {};
    fn non_async() = println!("hello");
    fn main() = non_async();
}
```

I expected to see this happen:
Ideally, this shouldn't be linted, but this should at least produce `clippy::manual_async_fn` with no suggestion.

Instead, this happened:
```
warning: this function can be simplified using the `async fn` syntax
  --> src/main.rs:12:13
   |
12 | /              $v $($($overrides)*)? fn $name$(<$($params)*>)?(
13 | |                  $($arg_name: $arg_typ),*
14 | |              ) $(-> $ret_ty)? {
15 | |                  $e
...  |
21 | |/ expr_fn! {
22 | ||     fn _hello() -> impl Future<Output = ()> = async {};
   | ||___________________________________________^
23 |  |     fn non_async() = println!("hello");
24 |  |     fn main() = non_async();
25 |  | }
   |  |_- in this macro invocation
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_async_fn
   = note: `#[warn(clippy::manual_async_fn)]` on by default
   = note: this warning originates in the macro `expr_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
help: make the function `async` and remove the return type
   |
12 ~             async $v $($($overrides)*)? fn $name$(<$($params)*>)?(
13 +                 $($arg_name: $arg_typ),*
14 +             ) $(-> $ret_ty)? {
15 +                 $e
16 +             }
17 +         )*
18 +     };
19 + }
20 + 
21 + expr_fn! {
22 ~     fn _hello() = async {};
   |
help: move the body of the async block to the enclosing function
   |
14 |             ) $(-> $ret_ty)? {}
   |                              ~~
```

Applying this suggestion results in:
```
error[E0752]: `main` function is not allowed to be `async`
  --> src/main.rs:12:13
   |
12 | /             $v async $($($overrides)*)? fn $name$(<$($params)*>)?(
13 | |                 $($arg_name: $arg_typ),*
14 | |             ) $(-> $ret_ty)? {
   | |_____________^ `main` function is not allowed to be `async`
...
21 | / expr_fn! {
22 | |     fn _hello() = {};
23 | |     fn non_async() = println!("hello");
24 | |     fn main() = non_async();
25 | | }
   | |_- in this macro invocation
   |
   = note: this error originates in the macro `expr_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Skipping the last change (replacing the function body in the macro definition) yields an additional error:
```
error[E0308]: mismatched types
  --> src/main.rs:24:17
   |
24 |     fn main() = non_async();
   |                 ^^^^^^^^^^^ expected `()`, found future
   |
note: calling an async function returns a future
  --> src/main.rs:24:17
   |
24 |     fn main() = non_async();
   |                 ^^^^^^^^^^^
help: consider `await`ing on the `Future`
   |
24 |     fn main() = non_async().await;
   |                            ++++++
help: consider using a semicolon here
   |
24 |     fn main() = non_async();;
   |                            +
```

### Version

```text
1.78.0-nightly (2024-03-02 5119208fd78a77547c70), on the Playground
```

### Additional Labels

@rustbot label +I-suggestion-causes-error +T-macros
